src/pbft/pbft.h

```diff
@@ -181,13 +185,36 @@ namespace pbft
         std::make_unique<ClientProxy<kv::TxHistory::RequestID, void>>(
           *message_receiver_base);
 
-      auto cb = [](Reply* m, void* ctx) {
+      auto reply_handler_cb = [](Reply* m, void* ctx) {
         auto cp =
           static_cast<ClientProxy<kv::TxHistory::RequestID, void>*>(ctx);
         cp->recv_reply(m);
       };
+      message_receiver_base->register_reply_handler(
+        reply_handler_cb, client_proxy.get());
 
-      message_receiver_base->register_reply_handler(cb, client_proxy.get());
+      auto batch_committed_cb =
+        [](uint64_t batch_id, void* batch, uint32_t size, void* ctx) {
+          auto ledger = static_cast<raft::LedgerEnclave*>(ctx);
+
+          struct LedgerEntry
+          {
+            uint64_t batch_id;
+            uint8_t data[0];
+          };
+
+          // TODO: This needs to be encrypted
+          std::vector<uint8_t> entry;
+          entry.resize(sizeof(LedgerEntry) + size);
+          auto ledger_entry = reinterpret_cast<LedgerEntry*>(entry.data());
+
+          ledger_entry->batch_id = batch_id;
+          memcpy(ledger_entry->data, batch, size);
+
+          ledger->put_entry(entry);
+        };
+      message_receiver_base->register_batch_committed(
+        batch_committed_cb, ledger.get());
```

Comment:
We talked about two modes of operation:
- one where we only signed checkpoints. In that mode, we should add checkpoints to the ledger but not pre-prepares.
- Another where we sign pre-prepares and prepares. In that one, we should add a signed pre-prepare to the ledger and we should add signed prepares for previous batches whose digests will be captured in the Merkle tree root that we are going to include in the signed pre-prepares.

Reply:
We are only running PBFT with large requests at the moment. We want to add these to the ledger. The pre-prepare includes their digests to determine an order for their execution. They should also be captured in the Merkle tree root. With BFT we do not need the write set in the ledger but we should have a hash of the write set to be able to check that execution was correct. We can also choose to have replies in the ledger. This can allow us to retransmit replies to old requests if ensuring excatly once semantics but it is not strictly necessary. 

For receipts to work, we want the pre-prepare and prepares to have signatures on hashes of the replies. If replies are in the ledger, we get that by just having the root of the merkle tree in pre-prepares. But it may be better to have a merkle tree per batch for replies to keep receipts shorter.

Also in the case of big (requests we want to add them to the ledger. The pre-prepare includes their digests

Reply:
The pre-prepares can be added to the ledger before sending them out. We should not wait for commit.

Reply:
@motdc at least in the case of governance transactions, I think it would be useful to keep write sets in the ledger because they make offline governance auditing considerably easier. What I think we want for offline governance verification is for members to be able to run very simple scripts (we already have an automated test along those lines), with which they can trace through governance decisions. If re-running the transactions is necessary, the process is slower and more involved (not to mention quite a bit more expensive to build and maintain).

This can hopefully be optional in the ledger, without impact on correctness or functionality (or even performance, considering governance is expected to be fairly infrequent).

Reply:
discussed offline - as per the discussion I am implementing option 2